File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:

    def getPermuta(self, arr, idx, ans):

        n = len(arr)

        if idx == n:
            ans.append(arr[:])   # copy array
            return

        for i in range(idx, n):

            arr[idx], arr[i] = arr[i], arr[idx]

            self.getPermuta(arr, idx + 1, ans)

            arr[idx], arr[i] = arr[i], arr[idx]   # backtracking


    def permute(self, nums):

        ans = []

        self.getPermuta(nums, 0, ans)

        return ans

File: Data Structures & Algorithms/rotate-array/submission-1.cpp
class Solution {
	public: void rotate(vector < int > & nums, int k) {
        ///// Time and Space O(n)
		// int n = nums.size();
		// vector < int > ans(n);
		// for(int i = 0; i < n; i++) {
		// 	ans[(i + k) % n] = nums[i];
		// }
		// nums = ans;




        /////// Time O(n) And Space is O(1)
        int n = nums.size();
        k %= n;
        reverse(nums.begin(), nums.end());
        reverse(nums.begin(), nums.begin() + k);
        reverse(nums.begin() + k, nums.end());
	}
};